File: Spring/Task 2 (Hierarchy)/sources/ice_cream/milky_ice_cream/milky_ice_cream.cpp
//
// Created by rinsler on 27.03.18.
//

#include "milky_ice_cream.h"


MilkyIceCream::MilkyIceCream(
        std::string name,
        std::string manufacturer,
        Ingredient flavor,
        Ingredient glaze,
        Ingredient syrup
) :
        IceCream(name, manufacturer),
        flavor(flavor),
        glaze(glaze),
        syrup(syrup)
{

}

std::string MilkyIceCream::getRecipe()
{
    std::string recipe =
            "Recipe of " + name +
            " from " + manufacturer +
            ": blend milk, egg yolk, cream, sugar and " +
            std::to_string(flavor.weight) + " grams of " + flavor.name +
            " and then whip and freeze it.";

    if (!syrup.name.empty()) {
        recipe += " Pour all with " + std::to_string(syrup.weight) + " grams of " + syrup.name + ".";
    }
    if (!glaze.name.empty()) {
        recipe += " Finally, cover product with " + std::to_string(glaze.weight) + " grams of " + glaze.name +
                  " and then immediately freeze it.";
    }

    return recipe;
}

File: Spring/Task 2 (IceCream)/IceCreamSources/IceCreamImplementations/MilkyIceCream.java
package IceCreamImplementations;

import com.sun.istack.internal.Nullable;
import IceCreamAbstract.IceCream;

public class MilkyIceCream extends IceCream {
    protected Ingredient flavor;
    protected Ingredient glaze;
    protected Ingredient syrup;

    public MilkyIceCream(String name, String manufacturer, Ingredient flavor, @Nullable Ingredient glaze, @Nullable Ingredient syrup) {
        super(name, manufacturer);
        this.flavor = flavor;
        this.glaze = glaze;
        this.syrup = syrup;
    }

    @Override
    public String getRecipe() {
        String recipe =
                "Recipe of " + name +
                        " from " + manufacturer +
                        ": blend milk, egg yolk, cream, sugar and " +
                        flavor.getWeight() + " grams of " + flavor.getName() +
                        " and then whip and freeze it.";

        if (syrup != null) {
            recipe += " Pour all with " + syrup.getWeight() + " grams of " + syrup.getName() + ".";
        }
        if (glaze != null) {
            recipe += " Finally, cover product with " + glaze.getWeight() + " grams of " + glaze.getName() +
                    " and then immediately freeze it.";
        }

        return recipe;
    }
}
